vyper/parser/context.py

```diff
@@ -104,28 +104,64 @@ def range_scope(self):
         yield
         self.in_range_expr = prev_value
 
-    def internal_memory_scope(self, scope_id):
-        # syntactic sugar for `make_blockscope` used to release
-        # memory after creating temporary internal variables
-        return self.make_blockscope(scope_id)
+    @contextlib.contextmanager
+    def internal_memory_scope(self):
+        """
+        Internal memory scope context manager.
+
+        Internal variables that are declared within this context are de-allocated
+        upon exitting the context.
+        """
+        scope_id = self._scope_id_iter
+        self._scope_id_iter += 1
+        self._scopes.add(scope_id)
+        yield
+
+        # Remove all variables that have specific scope_id attached
+        released = [
+            (k, v) for k, v in self.vars.items() if v.is_internal and scope_id in v.blockscopes
+        ]
+        for name, var in released:
+            self.memory_allocator.deallocate_memory(var.pos, var.size * 32)
+            del self.vars[name]
+
+        # Remove block scopes
+        self._scopes.remove(scope_id)
 
     @contextlib.contextmanager
-    def make_blockscope(self, blockscope_id):
-        self.blockscopes.add(blockscope_id)
+    def block_scope(self):
+        """
+        Block scope context manager.
+
+        All variables (public and internal) that are declared within this context
+        are de-allocated upon exiting the context.
+        """
+        scope_id = self._scope_id_iter
+        self._scope_id_iter += 1
+        self._scopes.add(scope_id)
         yield
 
-        # Remove all variables that have specific blockscope_id attached.
-        released = [(k, v) for k, v in self.vars.items() if blockscope_id in v.blockscopes]
+        # Remove all variables that have specific scope_id attached
+        released = [(k, v) for k, v in self.vars.items() if scope_id in v.blockscopes]
         for name, var in released:
             self.memory_allocator.deallocate_memory(var.pos, var.size * 32)
             del self.vars[name]
 
         # Remove block scopes
-        self.blockscopes.remove(blockscope_id)
+        self._scopes.remove(scope_id)
 
-    def _new_variable(self, name: str, typ: NodeType, var_pos: int) -> int:
+    def _new_variable(self, name: str, typ: NodeType, var_size: int, is_internal: bool) -> int:
+        if is_internal:
+            var_pos = self.memory_allocator.expand_memory(var_size)
```

Comment:
why do internal variables get a different allocation strategy?

Reply:
There are places in `parser` where an assumption is made that internal variables will be allocated sequentially within memory. If we don't do this, we end up with memory corruption.  I tried to unravel / refactor that logic but eventually decided it wasn't safe - I can't be sure I've found them all.

By always allocating new internal variables on the edge of memory like this, we don't risk memory corruption but still gain some efficiency by not expanding memory needlessly.  In the future when we refactor `parser` we should eliminate the assumption about how memory is sequenced, and then we can do away with this logic.

Reply:
weird. gotcha